[LLVM][DWARF] Refactor code for generating DWARF V5 .debug_names acce…

cmtice opened the pull request:
…lerator table.

Refactor the code that uniques the entries and computes the bucket count for the DWARF V5 .debug_names accelerator table. This will make it easier to share this bit of code wtih other tools such as LLD.

cmtice committed:
[LLVM][DWARF] Refactor code for generating DWARF V5 .debug_names accelerator table.

Refactor the code that uniques the entries and computes the bucket count
for the DWARF V5 .debug_names accelerator table. This will make it easier to
share this bit of code wtih other tools such as LLD.

felipepiovezan commented on llvm/include/llvm/ADT/STLExtras.h:
```diff
@@ -1994,6 +1994,10 @@ auto unique(Range &&R, Predicate P) {
 
+template <typename Range> auto unique(Range &&R) {
+  return std::unique(adl_begin(R), adl_end(R));
+}
```
I understand this is just moved code, but all functions in STLExtras have to be documented, even if they just defer to the stl documentation; see an example on the function just below this one

cmtice replied:
Done, in the new PR (requested by dwblaikie): https://github.com/llvm/llvm-project/pull/82312

felipepiovezan commented on llvm/include/llvm/BinaryFormat/Dwarf.h:
```diff
@@ -613,6 +613,24 @@ enum AcceleratorTable {
+// Uniquify the string hashes and calculate the bucket count for the
+// DWARF v5 Accelerator Table.
+inline uint32_t
+ComputeDebugNamesUniqueHashes(SmallVector<uint32_t, 0> &hashes) {
```
We avoid putting the SmallSize of SmallVectors in the interface, using SmallVectorImpl instead.
However, this function should just use a `MutableArrayRef<uint32_t>` instead, to communicate the fact that we are not inserting nor removing from the vector.

MaskRay replied:
New functions should use the `functionName` convention. (Older code may use `FunctionName`, which should be ignored.)

cmtice replied:
Done.

felipepiovezan replied:
I still see `SmallVector<uint32_t, 0>`? forgot to push maybe?

cmtice replied:
Whoops, you're right!  I've having a bit of trouble with my git push right now :-(  -- git is rejecting the push: "Updates were rejected because the remote contains work that you do not have locally..."  I tried a 'git merge' but that didn't fix the issue.  I hope I don't have to abandon this branch & PR and create a new one. If you know how to fix this, please let me know!  Otherwise, I'll keep you posted.

felipepiovezan replied:
If you show me the output of `git status` and the first few lines of `git log --oneline --graph` I might be able to tell you what's happening.

cmtice replied:
$ git status
On branch llvm-unique
Untracked files:
...
nothing added to commit but untracked files present (use "git add" to track)
$ git log --oneline --graph* 260397dcb8e4 (HEAD -> llvm-unique) [LLVM][DWARF] Refactor code for generating DWARF V4 .debug_names accelerator table
* 7f3980a7b2c9 (origin/main, origin/HEAD, main) [Fuzzer] Use user signal to coordinate handler shutdown (#82067)
* d2942a86d7b8 [MergeFunctions] Fix thunks for non-instruction debug info (#82080)
* 4c6043de0b83 [clang][InstallAPI] Add input file support to library (#81701)
* ae8facc1f8e8 [lldb-dap] Do not write over the existing error if launchCommands fail during debugger launch. (#82051)
* 4a23ab439cfb Fix Wcovered-switch-default Werror after #78252
* ef2837902212 [llvm-objcopy] Fix file offsets when PT_INTERP/PT_LOAD offsets are equal (#80562)
* 066773c41175 [X86] computeKnownBitsForTargetNode - add generic handling of PSHUFB
* 94100bc2fb1a [OpenMP][AIX]Add assembly file containing microtasking routines and unnamed common block definitions (#81770)
* 3b7ba2482e2c [C23] No longer assert on huge enumerator values (#81760)
* 9923d29cfa91 [VPlan] Merge main VPlan verifer with HCFG verifier.
* 9facaaddadd6 [ValueTracking] Improve tracking for constant range of `{s|u}rem C, x`
* aa7076fc9b7d [ValueTracking] Add tests for constant range of `{s|u}rem C, x`; NFC

...

cmtice replied:
I've given up and created a new pull request - it's simpler than fighting with git.

pogo59 replied:
Did you try to merge with the other branch where you did the ADT work? My guess is that the two branches had different bases, and if you merged without rebasing, your commits ended up buried in other commits from `main`.

felipepiovezan commented on llvm/include/llvm/BinaryFormat/Dwarf.h:
```diff
@@ -613,6 +613,24 @@ enum AcceleratorTable {
+ComputeDebugNamesUniqueHashes(SmallVector<uint32_t, 0> &hashes) {
+  uint32_t BucketCount = 0;
+
+  llvm::sort(hashes);
```
AFAICT we are already inside `namespace llvm`. (same comment below)

cmtice replied:
Fixed, thanks.

dwblaikie commented on llvm/include/llvm/ADT/STLExtras.h:
```diff
@@ -1994,6 +1994,10 @@ auto unique(Range &&R, Predicate P) {
   return std::unique(adl_begin(R), adl_end(R), P);
 }
 
+template <typename Range> auto unique(Range &&R) {
```
Please make this change (moving `unique(Range&&)` to STLExtras, and adding unit test coverage) in a separate commit (per https://llvm.org/docs/DeveloperPolicy.html#incremental-development )

cmtice replied:
Done: https://github.com/llvm/llvm-project/pull/82312

cmtice committed:
[LLVM][DWARF] Refactir code fir generating DWARF v5 .debug_names accelerator table.

Take the code for moving 'unique' from GenericUniformityImpl.h to
STLExtras.h out of the change (to put it into a separate change).

cmtice commented:
Note: This PR is now blocked by https://github.com/llvm/llvm-project/pull/82312; it won't compile properly until the other PR gets committed.

cmtice committed:
Merge branch 'main' into llvm-unique

cmtice commented:
I've gone ahead a created a new PR with all the latest updates addressing the review comments: https://github.com/llvm/llvm-project/pull/82394

cmtice commented:
Closing this pull request in favor of the new one.
